backend/src/resolvers.js

```diff
@@ -22,69 +24,127 @@ export class InvalidPasswordError extends UserInputError {
+      },
+      deletePost: async (_, { id }, context, info) => {
+        try {
+          const { authorId } = await context.dataSources.db.getPost(id)
+          if (authorId !== context.userId) return new DeletionOfOtherUsersPostForbiddenError()
+        } catch (error) {
+          if (error instanceof PostIdNotFoundError) return error
+          throw error
+        }
+
+        const post = await delegateQueryPost(id, context, info)
+        await context.dataSources.db.deletePost(id)
+        return post
+      },
+      upvotePost: async (_, { id }, context, info) => {
+        try {
+          await context.dataSources.db.upvotePost(id, context.userId)
+
+          return await delegateQueryPost(id, context, info)
+        } catch (error) {
+          if (error instanceof PostIdNotFoundError) return error
+          throw error
+        }
+      },
+      downvotePost: async (_, { id }, context, info) => {
+        try {
+          await context.dataSources.db.downvotePost(id, context.userId)
+
+          return await delegateQueryPost(id, context, info)
+        } catch (error) {
+          if (error instanceof PostIdNotFoundError) return error
+          throw error
+        }
+      },
+      signup: async (_, {
+        name,
+        email,
+        password
+      }, {
+        dataSources,
+        getUserAuthenticationToken
+      }) => {
+        if (password.length < 8) return new TooShortPasswordError()
+        const hasUserWithEmail = await dataSources.db.hasUserWithEmail(email)
+        if (hasUserWithEmail) return new EmailAlreadyExistsError(email)
+        const passwordHash = await bcrypt.hash(password, bcryptSaltRounds)
+        try {
+          const userId = await dataSources.db.createUser(name, email, passwordHash)
+          return getUserAuthenticationToken(userId)
+        } catch (error) {
+          if (error instanceof EmailAlreadyExistsError) return error
+          throw error
+        }
+      },
+      login: async (_, {
+        email,
+        password
+      }, {
+        dataSources,
+        getUserAuthenticationToken
+      }) => {
+        let user = null
+        try {
+          user = await dataSources.db.getUserByEmail(email)
+        } catch (error) {
+          if (error instanceof UserEmailNotFoundError) return error
+          throw error
+        }
+        return await bcrypt.compare(password, user.passwordHash)
+          ? getUserAuthenticationToken(user.id)
+          : new InvalidPasswordError()
       }
     },
-    downvotePost: (_, { id }, { dataSources, userId }) => {
-      try {
-        return dataSources.db.downvotePost(id, userId)
-      } catch (error) {
-        if (error instanceof PostIdNotFoundError) return error
-        throw error
+    Post: {
+      votes: {
+        selectionSet: '{ id }',
+        resolve: async (obj, _, { dataSources }) => {
+          const [upvotes, downvotes] = await dataSources.db.getVotesForPost(obj.id)
```

Comment:
I'm wondering whether the n+1 problem would occur here? Have you thought about it? I read that you can use [DataLoader](https://github.com/graphql/dataloader) to prevent it.

Reply:
I think that would send a request for every post requested.
It's not too bad, since the call uses [fast counts](https://neo4j.com/developer/kb/fast-counts-using-the-count-store/) to look up the counts in linear time (o(1)), but yes, it would be proper to add batching here!

Reply:
Okay, just wanted to clarify for my understanding. 😄